R/get-metadata.R

```diff
@@ -0,0 +1,106 @@
+#' Get task id values for a given round
+#'
+#' @param con a `hub_connection` class object.
+#' @param task_id Character string.  A task id name.
+#' @param round_id Character string.  A round id name. If rounds vary by a variable in hub,
+#'   argument is ignored and can be left `NULL` (default).
+#' @param flatten Logical. Whether to flatten required & optional task ids into a single
+#'  character vector.
+#'
+#' @return if `flatten = TRUE` (default), a character vector of task id values.
+#'   `flatten = FALSE`, a named list containing character vectors of
+#'   `required` and `optional` task id values.
+#' @export
+#' @family hub-metadata
+#' @examples
+#' con <- connect_hub(system.file("hub_1", package = "hubUtils"))
+#' get_task_id_vals(con, task_id = "location")
+#' con <- connect_hub(system.file("scnr_hub_1", package = "hubUtils"))
+#' get_task_id_vals(con, round_id = "round-1", task_id = "location")
+#' get_task_id_vals(con, round_id = "round-2", task_id = "age_group")
+#' get_task_id_vals(con,
+#'   round_id = "round-1", task_id = "location",
+#'   flatten = FALSE
+#' )
+get_task_id_vals <- function(con,
+                             task_id,
+                             round_id = NULL,
+                             flatten = TRUE) {
+  checkmate::assert_class(con, "hub_connection")
+  checkmate::assert_character(round_id, len = 1, null.ok = TRUE)
+  checkmate::assert_character(task_id, len = 1)
+
+
+  # validate inputs
+  round_id <- validate_round_ids(con, round_id)
+  # trigger validation error as a single task_id is being evaluated.
+  task_id <- validate_task_ids(con,
+    task_ids = task_id,
+    round_id = round_id, val_type = "error"
+  )
+
+
+  # extract task_id values from connection
+  values <- con[[round_id]]$model_tasks[[1]]$task_ids[[task_id]]
```

Comment:
I think that we don't want to subset to `model_tasks[[1]]` here. We may want to pull the values from all `model_tasks` entries.

Reply:
Could you clarify what you mean by:
> We may want to pull the values from all `model_tasks` entries.

In the currect hubmeta json files, the `model_tasks`element contains a list consisting of a single unnamed element which in turn contains two named list elements, `task_ids` & `output_types` (see below).

Hence I'm using `[[1]]` to access the underlying `task_ids` & `output_types`. Perhaps this superfluous unnamed list element shouldn't be there?

``` r
library(hubUtils)
con <- connect_hub(system.file("hub_1", package = "hubUtils")) 
round_id <- "round_id_from_variable"
con[[round_id]]$model_tasks[[1]]
#> $task_ids
#> $task_ids$origin_date
#> $task_ids$origin_date$required
#> NULL
#> 
#> $task_ids$origin_date$optional
#>  [1] "2022-01-08" "2022-01-15" "2022-01-22" "2022-01-29" "2022-02-05"
#>  [6] "2022-02-12" "2022-02-19" "2022-02-26" "2022-03-05" "2022-03-12"
#> [11] "2022-03-19" "2022-03-26" "2022-04-02" "2022-04-09" "2022-04-16"
#> [16] "2022-04-23" "2022-04-30" "2022-05-07" "2022-05-14" "2022-05-21"
#> [21] "2022-05-28" "2022-06-04" "2022-06-11" "2022-06-18"
#> 
#> 
#> $task_ids$location
#> $task_ids$location$required
#> NULL
#> 
#> $task_ids$location$optional
#>  [1] "01" "02" "04" "05" "06" "08" "09" "10" "11" "12" "13" "15" "16" "17" "18"
#> [16] "19" "20" "21" "22" "23" "24" "25" "26" "27" "28" "29" "30" "31" "32" "33"
#> [31] "34" "35" "36" "37" "38" "39" "40" "41" "42" "44" "45" "46" "47" "48" "49"
#> [46] "50" "51" "53" "54" "55" "56" "72" "78" "US"
#> 
#> 
#> $task_ids$horizon
#> $task_ids$horizon$required
#> NULL
#> 
#> $task_ids$horizon$optional
#> [1] 1 2 3 4
#> 
#> 
#> 
#> $output_types
#> $output_types$mean
#> $output_types$mean$type_id
#> $output_types$mean$type_id$required
#> NULL
#> 
#> $output_types$mean$type_id$optional
#> [1] NA
#> 
#> 
#> $output_types$mean$value
#> $output_types$mean$value$type
#> [1] "integer"
#> 
#> $output_types$mean$value$minimum
#> [1] 0
#> 
#> 
#> 
#> $output_types$quantile
#> $output_types$quantile$type_id
#> $output_types$quantile$type_id$required
#> NULL
#> 
#> $output_types$quantile$type_id$optional
#>  [1] 0.010 0.025 0.050 0.100 0.150 0.200 0.250 0.300 0.350 0.400 0.450 0.500
#> [13] 0.550 0.600 0.650 0.700 0.750 0.800 0.850 0.900 0.950 0.975 0.990
#> 
#> 
#> $output_types$quantile$value
#> $output_types$quantile$value$type
#> [1] "integer"
#> 
#> $output_types$quantile$value$minimum
#> [1] 0

str(con[[round_id]]$model_tasks[[1]])
#> List of 2
#>  $ task_ids    :List of 3
#>   ..$ origin_date:List of 2
#>   .. ..$ required: NULL
#>   .. ..$ optional: chr [1:24] "2022-01-08" "2022-01-15" "2022-01-22" "2022-01-29" ...
#>   ..$ location   :List of 2
#>   .. ..$ required: NULL
#>   .. ..$ optional: chr [1:54] "01" "02" "04" "05" ...
#>   ..$ horizon    :List of 2
#>   .. ..$ required: NULL
#>   .. ..$ optional: int [1:4] 1 2 3 4
#>  $ output_types:List of 2
#>   ..$ mean    :List of 2
#>   .. ..$ type_id:List of 2
#>   .. .. ..$ required: NULL
#>   .. .. ..$ optional: logi NA
#>   .. ..$ value  :List of 2
#>   .. .. ..$ type   : chr "integer"
#>   .. .. ..$ minimum: int 0
#>   ..$ quantile:List of 2
#>   .. ..$ type_id:List of 2
#>   .. .. ..$ required: NULL
#>   .. .. ..$ optional: num [1:23] 0.01 0.025 0.05 0.1 0.15 0.2 0.25 0.3 0.35 0.4 ...
#>   .. ..$ value  :List of 2
#>   .. .. ..$ type   : chr "integer"
#>   .. .. ..$ minimum: int 0
str(con[[round_id]]$model_tasks)
#> List of 1
#>  $ :List of 2
#>   ..$ task_ids    :List of 3
#>   .. ..$ origin_date:List of 2
#>   .. .. ..$ required: NULL
#>   .. .. ..$ optional: chr [1:24] "2022-01-08" "2022-01-15" "2022-01-22" "2022-01-29" ...
#>   .. ..$ location   :List of 2
#>   .. .. ..$ required: NULL
#>   .. .. ..$ optional: chr [1:54] "01" "02" "04" "05" ...
#>   .. ..$ horizon    :List of 2
#>   .. .. ..$ required: NULL
#>   .. .. ..$ optional: int [1:4] 1 2 3 4
#>   ..$ output_types:List of 2
#>   .. ..$ mean    :List of 2
#>   .. .. ..$ type_id:List of 2
#>   .. .. .. ..$ required: NULL
#>   .. .. .. ..$ optional: logi NA
#>   .. .. ..$ value  :List of 2
#>   .. .. .. ..$ type   : chr "integer"
#>   .. .. .. ..$ minimum: int 0
#>   .. ..$ quantile:List of 2
#>   .. .. ..$ type_id:List of 2
#>   .. .. .. ..$ required: NULL
#>   .. .. .. ..$ optional: num [1:23] 0.01 0.025 0.05 0.1 0.15 0.2 0.25 0.3 0.35 0.4 ...
#>   .. .. ..$ value  :List of 2
#>   .. .. .. ..$ type   : chr "integer"
#>   .. .. .. ..$ minimum: int 0
```

<sup>Created on 2022-10-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
Actually @elray1 ! I believe it is myself that has misunderstood here. When I was working on the JSON schema, I realised the simple example has **two** model task elements! Hence your comment is absolutely valid.

This needs a little more thought for how to handle so will work on it first thing next week.